datafusion/src/physical_plan/expressions/binary.rs

```diff
@@ -79,6 +76,193 @@ fn is_not_distinct_from_bool(
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn lt_eq_decimal_scalar(left: &DecimalArray, right: i128) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) <= right)?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn lt_eq_decimal(left: &DecimalArray, right: &DecimalArray) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) || right.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) <= right.value(i))?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn gt_decimal_scalar(left: &DecimalArray, right: i128) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) > right)?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn gt_decimal(left: &DecimalArray, right: &DecimalArray) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) || right.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) > right.value(i))?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn gt_eq_decimal_scalar(left: &DecimalArray, right: i128) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) >= right)?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn gt_eq_decimal(left: &DecimalArray, right: &DecimalArray) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) || right.is_null(i) {
+            bool_builder.append_null()?;
+        } else {
+            bool_builder.append_value(left.value(i) >= right.value(i))?;
+        }
+    }
+    Ok(bool_builder.finish())
+}
+
+fn is_distinct_from_decimal(
+    left: &DecimalArray,
+    right: &DecimalArray,
+) -> Result<BooleanArray> {
+    let mut bool_builder = BooleanBuilder::new(left.len());
+    for i in 0..left.len() {
+        if left.is_null(i) && right.is_null(i) {
+            bool_builder.append_value(false)?;
+        } else if left.is_null(i) || right.is_null(i) {
+            bool_builder.append_value(true)?;
+        } else {
+            bool_builder.append_value(left.value(i) != right.value(i))?;
```

Comment:
FWIW you can write this kind of structure in rust like


```suggestion
        match (left.is_null(i), right.is_null(i)) {
            (true, true) => bool_builder.append_value(false)?,
            (true, false) | (false | true) => bool_builder.append_value(true)?;
            _ => bool_builder.append_value(left.value(i) != right.value(i))?;
        }
```

Where the benefit is that the compiler will ensure you all your branches are reachable and you have covered all the cases. 

I don't think this needs to be changed, I just wanted to point it out while I was working through this PR

Reply:
it's great and it will be resolved in https://github.com/apache/arrow-datafusion/pull/1554/files